src/game/connect4.rs

```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
             return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
                 self.state,
             )));
+        }
+        let request_payload = serde_json::from_value::<Connect4RequestPayload>(game_move.payload)?;
+        let column = request_payload.column;
+        let mut player = match self.game.turn {
+            Token::Red => self.players.get(0).unwrap().clone(),
+            Token::Blue => self.players.get(1).unwrap().clone(),
         };
-        todo!()
+        let mut user = game_move.player;
+        if player == user {
+            self.game.moves(column)?;
+            let win = self.game.winning_move(column);
+            let draw = self.game.draw_move();
+            if win {
+                let mut winner = match self.game.turn {
+                    Token::Red => self.players.get(0).unwrap().clone(),
+                    Token::Blue => self.players.get(1).unwrap().clone(),
+                };
+                self.winner.push(winner);
+            }
+            if win || draw {
+                self.game.completed = true;
+            } else {
+                self.game.switch_token();
+            }
+            Ok(())
+        } else {
+            Err(actix_web::Error::from(
+                GameAdapterError::WrongPlayerRequest(user), // return the one who made the request
+            ))
+        }
     }
 
     fn get_encoded_state(&self) -> actix_web::Result<GenericGameState> {
+        let encoded_board = self
+            .game
+            .board
+            .iter()
+            .map(|col| {
+                col.iter()
+                    .map(|&token| match token {
+                        Token::Red => self.players.get(0),
+                        Token::Blue => self.players.get(1),
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
             can_move: vec![self.next_move.clone()],
             winners: vec![],
-            payload: serde_json::to_value(&self.game.board)?,
+            payload: serde_json::to_value(&encoded_board)?,
         })
     }
 }
 
-impl Connect4<'_> {}
+impl Connect4 {
+    fn get_cell_at(&self, row: usize, col: usize) -> Option<Token> {
+        Some(*self.board.get(col)?.get(row)?)
+    }
+
+    fn insert_move_if_legal(&mut self, column: usize) -> actix_web::Result<()> {
+        if column >= COL_SIZE {
+            return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
```

Comment:
Is this really an "invalid game state" error? The game state "in progress" is valid - it's the *move* that is invalid.